word/pkg/graph/docs.go

```diff
@@ -17,13 +24,220 @@ func (d DocInfo) String() string {
+		// Try to get the folder by path.
+		item, err := getItemByPath(ctx, client, driveID, currentPath)
+		if err != nil {
+			if !strings.Contains(err.Error(), "item not found") {
+				return nil, fmt.Errorf("failed to get item by path %q: %w", currentPath, err)
+			}
+			// Assume an error indicates the folder was not found.
+			// Create the folder in the current parent folder.
+			newFolder := graphmodels.NewDriveItem()
+			newFolder.SetName(&part)
+			// Mark the item as a folder.
+			newFolder.SetFolder(graphmodels.NewFolder())
+			// Set conflict behavior to "rename" (or "fail") to avoid naming conflicts.
+			newFolder.SetAdditionalData(map[string]any{
+				"@microsoft.graph.conflictBehavior": "fail",
+			})
+			createdFolder, err := client.Drives().
+				ByDriveId(driveID).
+				Items().
+				ByDriveItemId(currentFolderID).
+				Children().
+				Post(ctx, newFolder, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create folder %q: %w", part, err)
+			}
+			currentItem = createdFolder
+		} else {
+			// Folder already exists.
+			currentItem = item
+		}
+		// Update the parent folder for the next iteration.
+		currentFolderID = deref(currentItem.GetId())
+	}
+
+	return currentItem, nil
+}
+
+// uploadFileContent uploads file content as a new drive item under the specified parent folder.
+func uploadFileContent(ctx context.Context, client *msgraphsdkgo.GraphServiceClient, driveID string, parentID string, filename string, content []byte) (graphmodels.DriveItemable, error) {
+	if parentID == "" {
+		parentID = "root"
+	}
+
+	var doc graphmodels.DriveItemable
+	var err error
+
+	// Check if the file exists by path or ID
+	if strings.ContainsAny(filename, "/.:_-") {
+		doc, err = getItemByPath(ctx, client, driveID, filename)
+		if err != nil {
+			slog.Info("Failed to get item by path. It may not exist so we create it.", "path", filename, "error", err)
+		}
+	} else {
+		doc, err = client.Drives().ByDriveId(driveID).Items().ByDriveItemId(filename).Get(ctx, nil)
+		if err != nil {
+			slog.Info("Failed to get item by ID. It may not exist so we create it.", "name", filename, "error", err)
+		}
+	}
+
+	// Build the URL for a simple upload:
+	// PUT /drives/{drive-id}/items/{parent-id}:/{filename}:/content
+	requestInfo := kiota.NewRequestInformation()
+	requestInfo.PathParameters = map[string]string{
+		"baseurl": client.RequestAdapter.GetBaseUrl(), // for some weird reason, this is deprecated, but also the only way to get it working
+	}
+	requestInfo.Method = kiota.PUT
+	requestInfo.SetStreamContentAndContentType(content, "text/plain")
+
+	if doc == nil {
+		slog.Info("File does not exist. Creating.", "name", filename)
+		requestInfo.UrlTemplate = "{+baseurl}/drives/{driveid}/items/{parentid}:/{filename}:/content"
+		requestInfo.PathParametersAny = map[string]any{
+			"driveid":  driveID,
+			"parentid": parentID,
+			// URL-encode the filename if necessary.
+			"filename": filename,
```

Comment:
Doesn't look like this is getting URL-encoded?

Reply:
Correct, that broke everything so I removed it :D